# Solanum Dataset Metrics

We wish to obtain the following information for each file:

1. Section to which it belongs
1. Species
1. Source of file
1. Size (MB)
1. Resolution
1. Type of file
1. hash (fingerprint)

## Image fingeprinting
We'll also use the `ImageHash` library to calculate the fingerprint of the images in order to identify near-duplicates.

## References:
1. https://realpython.com/fingerprinting-images-for-near-duplicate-detection/
1. https://github.com/JohannesBuchner/imagehash
1. https://www.hackerfactor.com/blog/index.php?/archives/529-Kind-of-Like-That.html


# Imports

In [ ]:
!pip install imagehash pillow pandas numpy matplotlib

In [2]:
import os

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
import imagehash

pd.set_option('max_rows', 200)
pd.set_option('max_colwidth', 400)

In [7]:
#!echo %cd%
!echo "$PWD"

/mnt/pepper/CODE/ITESM/tesis-dataset-downloader


In [8]:
#root_folder = "P:/CODE/ITESM/tesis-dataset-downloader/solanum_output"
root_folder = "/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/solanum_output"

# Tools

In [9]:
def plot_histogram(df: pd.DataFrame, nrows: int, ncols: int):
    """
    Plots a series of histograms for a dataframe in a nrows x ncols distribution
    """
    fig, ax = plt.subplots(nrows=nrows,
                           ncols=ncols,
                           figsize=(14,14))
    
    for n, col in enumerate(df.columns):
        plt.subplot(nrows, ncols, n+1)
        df[col].plot.hist(ax = plt.gca(),
                          bins=20,
                          legend=True)

# 1 Calculate image attributes and fingerprints

dHash is the chosen fingerprint algorithm to use. Like aHash and pHash, dHash is pretty simple to implement and is far more accurate than it has any right to be. As an implementation, dHash is nearly identical to aHash but it performs much better. While aHash focuses on average values and pHash evaluates frequency patterns, dHash tracks gradients.

In [ ]:
records = []

for root, dirs, files in os.walk(root_folder):
    
    if root == root_folder:
        continue
    #print(f"Root: {root}")
    #print(f"Dirs: {dirs}")
    #print(f"Files: {files}")
    #print("-" * 50)

    print(f"Reading files in {root}...")

    for f in files:
        full_path = os.path.join(root, f)
        
        # Section to which it belongs
        # Species
        # Source of file
        section, species, _, source, _ = f.split("_")

        # Type of file
        file_extension = os.path.splitext(full_path)[1].replace(".", "").lower()

        # Normalize JPG extensions
        file_extension = "jpg" if file_extension == "jpeg" else file_extension

        # Size (MB)
        filesize_mb = round(os.path.getsize(full_path) / 1024 / 1024, 4)

        # Resolution
        image = Image.open(full_path)
        width, height = image.size

        # fingerprint
        hash = str(imagehash.dhash(image))
        
        row = {
            "section": section,
            "species": species,
            "filesize_mb": filesize_mb,
            "width": width,
            "height": height,
            "image_type": file_extension,
            "source": source,
            "hash": hash,
            "full_path": full_path,
            "keep": 1
        }

        records.append(row)

Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\anarrhichomenum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\basarthrum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\brevantherum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\crinitum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\dulcamara...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\erythrotrichum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\gonatotrichum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\herposolanum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\herpystrichum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\holophylla...


p:\code\itesm\tesis-dataset-downloader\venv-tesis\lib\site-packages\PIL\TiffImagePlugin.py:793: UserWarning: Corrupt EXIF data.  Expecting to read 12 bytes but only got 10. 
  warnings.warn(str(msg))


Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\lasiocarpa...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\lycopersicon...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\micracantha...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\pachyphylla...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\periscariae...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\pteroidea...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\solanum...
Reading files in P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva...


In [ ]:
report = pd.DataFrame(records)

display(report.head(25))


section         species  filesize_mb  width  height image_type  \
0   acanthophora     acerifolium       8.9974   3750    5500        jpg   
1   acanthophora     acerifolium       7.3703   3750    5500        jpg   
2   acanthophora     acerifolium       7.0359   3750    5500        jpg   
3   acanthophora     acerifolium       7.8730   3750    5500        jpg   
4   acanthophora     acerifolium       5.9595   3750    5500        jpg   
5   acanthophora     acerifolium       6.3095   3750    5500        jpg   
6   acanthophora     acerifolium       6.4096   3750    5500        jpg   
7   acanthophora     acerifolium       9.5827   6738    8934        jpg   
8   acanthophora     acerifolium       4.3827   3231    5008        jpg   
9   acanthophora     acerifolium       5.7148   3260    5011        jpg   
10  acanthophora     acerifolium      11.0593   6708    8930        jpg   
11  acanthophora     acerifolium      10.4834   6740    8943        jpg   
12  acanthophora  aculeatissimum       3.9350   3456    5184        jpg   
13  acanthophora     capsicoides       8.2127   3750    5500        jpg   
14  acanthophora     capsicoides       7.4239   3750    5500        jpg   
15  acanthophora        mammosum       0.6219    871    1440        jpg   
16  acanthophora        mammosum       0.5724    870    1440        jpg   
17  acanthophora        mammosum       0.7719    869    1440        jpg   
18  acanthophora        mammosum       6.9009   3750    5500        jpg   
19  acanthophora        mammosum       6.7975   3750    5500        jpg   
20  acanthophora        mammosum       5.9054   3750    5500        jpg   
21  acanthophora        mammosum       0.1026   1004    1520        jpg   
22  acanthophora        mammosum      11.3181   6717    9001        jpg   
23  acanthophora        mammosum       9.9155   6749    8957        jpg   
24  acanthophora        mammosum       7.9070   6733    8931        jpg   

     source              hash  \
0      gbif  8c86b6326a6e0cca   
1      gbif  bea48d0a2aac92b2   
2      gbif  b893b2a893b6a2e6   
3      gbif  98a6963cb69a8aca   
4      gbif  96963696ca8adaf2   
5      gbif  b4938b8682beb2f2   
6      gbif  96a893832cb2a2c2   
7   idigbio  e6e62e3cc6879551   
8      gbif  8c0e26574accce88   
9      gbif  36379b933e8eccac   
10  idigbio  343ec6a6262f1d15   
11  idigbio  da22266c26371999   
12     gbif  6736318f8fb0c1c8   
13     gbif  b08ace222696a6aa   
14     gbif  8e1a3abe1626ead2   
15     gbif  f43f3b3331b1ec30   
16     gbif  ce23697a3a33a333   
17     gbif  39192d6d891fc622   
18     gbif  8aa29aa6323a92e6   
19     gbif  8ea2998e8e3a32a6   
20     gbif  8ab2b6caaaaccaf2   
21     gbif  92b2b5979686e8a2   
22  idigbio  06323e2e26279dd9   
23  idigbio  9896623a0c8db1a9   
24  idigbio  f0da8e9292d7e1d9   

                                                                                                                                             full_path  \
0                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928496814_gbif_2700.jpg   
1                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928638909_gbif_2609.jpg   
2                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928803178_gbif_2510.jpg   
3                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1929578053_gbif_2035.jpg   
4                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1929578350_gbif_2034.jpg   
5                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1930342960_gbif_1533.jpg   
6                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifo

In [ ]:
report.to_csv(os.path.join(root_folder, "Downloaded_images_report.csv"))

# 2 Detecting and removing duplicates

In [ ]:
#report = pd.read_csv(os.path.join(root_folder, "Downloaded_images_report.csv"))
print(report.shape)

(8937, 11)


In [ ]:
duplicates = report[report["hash"].duplicated(keep=False)].sort_values("hash")

# All duplicates will be kept until we mark them for deletion
duplicates["keep"] = 1

display(duplicates)
print(duplicates.shape)

Unnamed: 0        section          species  filesize_mb  width  height  \
2844        2844     holophylla   pseudocapsicum       0.0387    497     768   
2746        2746     holophylla   pseudocapsicum       0.0387    497     768   
5749        5749         petota         edinense       0.0697    496     768   
5720        5720         petota         edinense       0.0697    496     768   
5745        5745         petota         edinense       0.0610    493     768   
5717        5717         petota         edinense       0.0610    493     768   
8196        8196          torva      lanceolatum       1.0476   2048    3362   
8531        8531          torva      lanceolatum       1.0249   2592    4255   
3110        3110     holophylla        pubigerum       5.7115   3228    5026   
3060        3060     holophylla        pubigerum       5.7115   3228    5026   
2843        2843     holophylla   pseudocapsicum       0.0468    466     768   
2745        2745     holophylla   pseudocapsicum       0.0468    466     768   
5247        5247         petota    bulbocastanum       0.0360    496     768   
5246        5246         petota    bulbocastanum       0.0360    496     768   
5384        5384         petota    cardiophyllum       0.0448    496     768   
5481        5481         petota    cardiophyllum       0.0448    496     768   
1580        1580   herposolanum      glaucescens       0.0366    490     768   
1632        1632   herposolanum      glaucescens       0.0366    490     768   
6903        6903         petota       verrucosum       1.2977   2753    4359   
6866        6866         petota       verrucosum       1.1759   2048    3243   
2740        2740     holophylla   pseudocapsicum       0.0417    494     768   
2830        2830     holophylla   pseudocapsicum       0.0417    494     768   
3754        3754      melongena    angustifolium       2.2092   1300    2000   
3753        3753      melongena    angustifolium       2.2089   1300    2000   
6792        6792         petota           tarnii       0.0565    491     768   
6801        6801         petota           tarnii       0.0565    491     768   
6357        6357         petota     stoloniferum       1.2167   2048    2999   
6774        6774         petota     stoloniferum       1.5163   2977    4359   
8592        8592          torva      lanceolatum       2.3620   1419    2000   
8587        8587          torva      lanceolatum       2.3623   1419    2000   
5510        5510         petota    cardiophyllum       0.0341    494     768   
5385        5385         petota    cardiophyllum       0.0341    494     768   
3537        3537      melongena    angustifolium       0.5252    951    1440   
4923        4923      melongena        rostratum       0.5252    951    1440   
1892        1892     holophylla         aligerum       0.0597    496     768   
1891        1891     holophylla         aligerum       0.0597    496     768   
1821        1821     holophylla         aligerum       0.0597    496     768   
1820        1820     holophylla         aligerum       0.0597    496     768   
2238        2238     holophylla        nigricans       0.0577    455     768   
2407        2407     holophylla        nigricans       0.0577    455     768   
2239        2239     holophylla        nigricans       0.0487    483     768   
2408        2408     holophylla        nigricans       0.0487    483     768   
6523        6523         petota     stoloniferum       0.0478    491     768   
6305        6305         petota     stoloniferum       0.0478    491     768   
5364        5364         petota    bulbocastanum       2.0898   1383    2000   
5365        5365         petota    bulbocastanum       2.0901   1383    2000   
8409        8409          torva      lanceolatum       0.0466    490     768   
8162        8162          torva      lanceolatum       0.0466    490     768   
1519        1519  gonatotrichum         deflexum       2.5137   1738    2000   
1561  

(190, 11)


In [ ]:
dups_count_by_hash = duplicates.groupby(["hash"]).size().reset_index(name='count').sort_values(["count", "hash"], ascending=False)
display(dups_count_by_hash)
print(dups_count_by_hash.shape)

hash  count
33  7476646c59572d39      6
17  451b169491cddd73      4
91  f89ab23c53935bd8      2
90  e9299696cae28c9b      2
89  e6c74f3e341e1b71      2
88  e36e0c0c183121b1      2
87  e22969c24c642663      2
86  ddf526363623e5e5      2
85  dabaca143636ece8      2
84  d699093e18ca0312      2
83  d5af66421a894d4d      2
82  d31b2bdcb0e9f170      2
81  d29e2e368f1336c8      2
80  d1990e361e0b3727      2
79  d163279a8e94db66      2
78  d0c3d77672b32160      2
77  d00634690d134b09      2
76  cec5c5a796ac446c      2
75  caeac8eab2b0ae8c      2
74  c94533db8cc6d9d9      2
73  c9012b345a121a5b      2
72  c88e86cc9998e0e0      2
71  c8229a4729b2e6e8      2
70  c78969323438c8f1      2
69  c6c3c3c1c927372f      2
68  c69edc8c94999890      2
67  c40cf28361659549      2
66  baa0a48ca6dab2a2      2
65  b6a5848691a0a4e8      2
64  b2326c6a2ed8f180      2
63  b1914eaccc9a9e92      2
62  b18fbda12f6b3b63      2
61  b09a368e19616810      2
60  b02e4fcd47177658      2
59  b02c2e30b0a9f090      2
58  ac2c3e96369698e0      2
57  aa929a6c66a498da      2
56  a8be968a82969ae8      2
55  a6967858694e4fd9      2
54  a34f23271e373589      2
53  a2929cce969c8aa6      2
52  9e9a323b3a9579e8      2
51  9e8eb09aa6f6929a      2
50  9e8e362e2f2b33e8      2
49  9a9adece8ab2aab2      2
48  98a7b39a9e9ac2e8      2
47  9898a8b2b2acecb0      2
46  9474188a939aa6a4      2
45  929eaea28292a2aa      2
44  92969a899b9e26c8      2
43  9199333838946569      2
42  90959b89919b2664      2
41  8e8e16176313cad8      2
40  8cbabab69aa4aefa      2
39  8c9aa3a3e35bcef2      2
38  8c91982e6696b2aa      2
37  8c4e4c5a4e7cb0c8      2
36  8c4e4c584f49787c      2
35  7c7cc738343606ed      2
34  74cacb33d60e0696      2
32  6d65919b9d95d633      2
31  6c248c4e5a868698      2
30  6b651ac76d29272c      2
29  646c06ced9c9df9c      2
28  642c06cc8e9e9ee8      2
27  61594c5c70efe4d2      2
26  606cccc68c88c9f1      2
25  6065616436344949      2
24  604a4a7aeae8f2c8      2
23  53617b41773e1a0e      2
22  4ece1216632b034e      2
21  4d51d31a4a4e65a7      2
20  49273d0dc83381c4      2
19  46563665335e7eac      2
18  46563625339eb6e8      2
16  43e7d1d3c20381d2      2
15  4301d1c1c1e11b1b      2
14  3b876634135beded      2
13  3ad83a4a3b0f2549      2
12  394b221a163c6773      2
11  39319561486a6a6e      2
10  39293e3616379c6a      2
9   3877679a86c988b0      2
8   37350526932929c1      2
7   35350d1741638d8d      2
6   313159242c281969      2
5   2c3e2e2c3e3c32e8      2
4   2c36672b3ae4b882      2
3   288c363698909090      2
2   272f2d4dc3f32d0d      2
1   2325954e06898916      2
0   0939260d383c2478      2

(92, 2)


We can see that we have 190 duplicated records out of which only 92 are unique images, many of them of the same size and resolution, some have different size but represent the same image and some of them are classified as being a member of a section and have the same picture classified as another, we need to remove these and make sure that the pictures are actually in the same place.

To break a tie and see what records we are keeping we will:
1. Say that absolute duplicates are ones such that they have the same:
    * hash
    * section
    * species
    * filesize_mb

   For these, it doesn't matter which ones we keep, let's keep one record and remove the rest.
1. For the records that are still duplicated on their `hash` but have different sizes, we'll set the `keep` column of the rows with the smallest size to 0 so that we can remove them.
1. Remove those records that have `keep = 0`. Out of those we will again get a report of the records that still have the same `hash`. These are the records that may have different section or species classification, for these we will determine which ones to keep by asking a biologist to make the final determination of which records are correct and which aren't, we'll manually set the `keep` value to 0 for those we want removed.
1. Remove the records that have a `keep = 0`.
1. Remove the `keep` columns.

This processing should leave us only with the best resolutions for the specimens and the correct images to be preprocessed for feature extraction.

## 2.1 Removal of absolute duplicates

In [ ]:
absolute_columns = ["hash", "filesize_mb", "section", "species"]
duplicates.loc[duplicates.duplicated(subset=absolute_columns, keep="first"), "keep"] = 0

In [ ]:
duplicates.sort_values(["hash", "keep"], ascending=False)

Unnamed: 0        section          species  filesize_mb  width  height  \
8416        8416          torva      lanceolatum       0.0458    468     768   
8134        8134          torva      lanceolatum       0.0458    468     768   
2213        2213     holophylla      malacothrix       0.0362    489     768   
2197        2197     holophylla      malacothrix       0.0362    489     768   
6008        6008         petota        oxycarpum       1.3979   2048    3235   
6036        6036         petota        oxycarpum       1.5676   2746    4338   
6192        6192         petota  stenophyllidium       0.0415    488     768   
6195        6195         petota  stenophyllidium       0.0415    488     768   
6724        6724         petota     stoloniferum       1.7507   1248    2000   
6718        6718         petota     stoloniferum       1.7504   1248    2000   
5798        5798         petota       hjertingii       1.2235   2048    2953   
5820        5820         petota       hjertingii       1.5602   2800    4037   
6295        6295         petota     stoloniferum       0.0535    465     768   
6530        6530         petota     stoloniferum       0.0535    465     768   
70            70   acanthophora     myriacanthum       0.0434    504     768   
31            31   acanthophora     myriacanthum       0.0434    504     768   
5694        5694         petota         demissum       0.0648    500     768   
5527        5527         petota         demissum       0.0648    500     768   
5860        5860         petota        iopetalum       0.0425    500     768   
5930        5930         petota        iopetalum       0.0425    500     768   
3343        3343     lasiocarpa         candidum       0.0428    465     768   
3272        3272     lasiocarpa         candidum       0.0428    465     768   
3344        3344     lasiocarpa         candidum       0.0361    496     768   
3273        3273     lasiocarpa         candidum       0.0361    496     768   
1002        1002   brevantherum       umbellatum       0.0348    482     768   
839          839   brevantherum       umbellatum       0.0348    482     768   
6033        6033         petota        oxycarpum       1.4370   2746    4311   
6007        6007         petota        oxycarpum       1.2624   2048    3215   
7347        7347        solanum       nigrescens       0.1731    501     768   
7487        7487        solanum       nigrescens       0.1731    501     768   
7764        7764          torva       dasyanthum       1.2286   2048    3234   
7778        7778          torva       dasyanthum       1.5714   2800    4422   
4636        4636      melongena        houstonii       8.5188   3750    5500   
4642        4642      melongena        houstonii       8.5390   3750    5500   
6115        6115         petota      polyadenium       0.0551    493     768   
6082        6082         petota      polyadenium       0.0551    493     768   
8417        8417          torva      lanceolatum       0.0384    499     768   
8135        8135          torva      lanceolatum       0.0384    499     768   
6853        6853         petota       verrucosum       1.1550   2753    4359   
6867        6867         petota       verrucosum       1.0611   2048    3243   
4773        4773      melongena        houstonii       0.0545    467     768   
5033        5033    micracantha       tampicense       0.0545    467     768   
6279        6279         petota     stoloniferum       1.1599   2732    4373   
6350        6350         petota     stoloniferum       1.0491   2048    3278   
6289        6289         petota     stoloniferum       0.0458    500     768   
6547        6547         petota     stoloniferum       0.0458    500     768   
6472        6472         petota     stoloniferum       1.0026   2725    4345   
6358        6358         petota     stoloniferum       0.9231   2048    3266   
5953        5953         petota      michoacanum       0.0442    496     768   
5961  

## 2.2 Pruning of smaller pictures that are have the same hash but different sizes

In [ ]:
ordered_dups = duplicates[duplicates["keep"] == 1].sort_values(["hash", "filesize_mb"], ascending=False)
ordered_dups.loc[ordered_dups.duplicated(subset=["hash", "section", "species"], keep="first"), "keep"] = 0

In [ ]:
# Update the original duplicates dataframe
duplicates.loc[duplicates.index.isin(ordered_dups[ordered_dups["keep"] == 0].index), "keep"] = 0

In [ ]:
print(ordered_dups[ordered_dups["keep"] == 0].shape)
print(duplicates[duplicates["keep"] == 0].shape)

(27, 11)
(93, 11)


## 2.3 Pruning of duplicated records that are classified in multiple sections/species

In [ ]:
reclassification_dups = ordered_dups[ordered_dups["keep"] == 1].drop("Unnamed: 0", axis=1)
reclassification_dups[reclassification_dups.duplicated("hash", keep=False)]

section        species  filesize_mb  width  height image_type  \
4773    melongena      houstonii       0.0545    467     768        jpg   
5033  micracantha     tampicense       0.0545    467     768        jpg   
5786       petota       hintonii       2.8609   7079   10079        jpg   
6105       petota    polyadenium       2.8609   7079   10079        jpg   
5886       petota      iopetalum       1.6025   2732    4359        jpg   
6006       petota      oxycarpum       1.4398   2048    3268        jpg   
4774    melongena      houstonii       0.0425    487     768        jpg   
5032  micracantha     tampicense       0.0425    487     768        jpg   
3537    melongena  angustifolium       0.5252    951    1440        jpg   
4923    melongena      rostratum       0.5252    951    1440        jpg   

       source              hash  \
4773  idigbio  c8229a4729b2e6e8   
5033     gbif  c8229a4729b2e6e8   
5786     gbif  b1914eaccc9a9e92   
6105     gbif  b1914eaccc9a9e92   
5886  idigbio  b02e4fcd47177658   
6006     gbif  b02e4fcd47177658   
4774  idigbio  6b651ac76d29272c   
5032     gbif  6b651ac76d29272c   
3537     gbif  43e7d1d3c20381d2   
4923  idigbio  43e7d1d3c20381d2   

                                                                                                                                       full_path  \
4773  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena\melongena_houstonii_a25edb3c-eb4f-40eb-a9a9-2aa00578056f_idigbio_5006.jpeg   
5033                          P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\micracantha\micracantha_tampicense_1055692588_gbif_4898.jpeg   
5786                                       P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_hintonii_2430227118_gbif_420.jpeg   
6105                                    P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_polyadenium_2430448876_gbif_407.jpeg   
5886          P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_iopetalum_4bff8a9b-552c-482a-9548-cc175d4d5802_idigbio_948.jpg   
6006                                       P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_oxycarpum_144897192_gbif_6725.jpg   
4774  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena\melongena_houstonii_a25edb3c-eb4f-40eb-a9a9-2aa00578056f_idigbio_5007.jpeg   
5032                          P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\micracantha\micracantha_tampicense_1055692588_gbif_4897.jpeg   
3537                           P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena\melongena_angustifolium_1135482425_gbif_4313.jpeg   
4923  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena\melongena_rostratum_eac14d4c-209b-496a-a432-2efef51fb50b_idigbio_3480.jpeg   

      keep  
4773     1  
5033     1  
5786     1  
6105     1  
5886     1  
6006     1  
4774     1  
5032     1  
3537     1  
4923     1

We need to consult with a taxonomist to get the records that must be kept and which need to be removed from the dataset.

This error of having the same picture in two different species or even sections most likely happened because of capturing errors or determination errors made by the taxonomist that uploaded the specimen to the online database.

It is noticed that the picture identified with hash `c8229a4729b2e6e8` and `6b651ac76d29272c` represent the same specimen but substantial changes were made to the specimen sheet.

In [ ]:
indices_to_delete = [5033,
                     6105,
                     5886,
                     4774,5032,
                     3537
                     ]
reclassification_dups.loc[reclassification_dups.index.isin(indices_to_delete), "keep"] = 0



In [ ]:
# Update the original duplicates dataframe
duplicates.loc[duplicates.index.isin(reclassification_dups[reclassification_dups["keep"] == 0].index), "keep"] = 0

In [ ]:
# Update the original report to remove all duplicates
report.loc[report.index.isin(duplicates[duplicates["keep"] == 0].index), "keep"] = 0

In [ ]:
dedup_images = report[report["keep"] == 1].drop(["keep", "Unnamed: 0"], axis=1)
dedup_images.to_csv(os.path.join(root_folder, "Downloaded_dedup_images_report.csv"))

# 3 Metrics on deduplicated images

In [ ]:
dedup_images = pd.read_csv(os.path.join(root_folder, "Downloaded_dedup_images_report.csv"))
dedup_images.describe(include="all")

Unnamed: 0 section      species  filesize_mb        width  \
count   8838.000000    8838         8838  8838.000000  8838.000000   
unique          NaN      21          127          NaN          NaN   
top             NaN  petota  lanceolatum          NaN          NaN   
freq            NaN    1734          533          NaN          NaN   
mean    4460.286264     NaN          NaN     6.191186  3640.765897   
std     2582.206533     NaN          NaN     8.051593  2122.252540   
min        0.000000     NaN          NaN     0.006400   120.000000   
25%     2224.250000     NaN          NaN     1.861500  1216.000000   
50%     4451.500000     NaN          NaN     5.873050  3750.000000   
75%     6700.750000     NaN          NaN     8.962625  6023.250000   
max     8936.000000     NaN          NaN    96.622100  7319.000000   

              height image_type source              hash  \
count    8838.000000       8838   8838              8838   
unique           NaN          1      2              8838   
top              NaN        jpg   gbif  8aadb6828a12aa8c   
freq             NaN       8838   6314                 1   
mean     5148.799050        NaN    NaN               NaN   
std      2743.151575        NaN    NaN               NaN   
min       160.000000        NaN    NaN               NaN   
25%      2000.000000        NaN    NaN               NaN   
50%      5500.000000        NaN    NaN               NaN   
75%      8679.250000        NaN    NaN               NaN   
max     10319.000000        NaN    NaN               NaN   

                                                                                                                                   full_path  
count                                                                                                                                   8838  
unique                                                                                                                                  8838  
top     P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva\torva_ferrugineum_b3b3c31f-accb-4821-9838-76841cbe14e6_idigbio_2464.jpeg  
freq                                                                                                                                       1  
mean                                                                                                                                     NaN  
std                                                                                                                                      NaN  
min                                                                                                                                      NaN  
25%                                                                                                                                      NaN  
50%                                                                                                                                      NaN  
75%                                                                                                                                      NaN  
max                                                                                                                                      NaN

## Count per section

In [ ]:
count_per_section = dedup_images.groupby(["section"]).size().reset_index(name='count').sort_values("count", ascending=False)
display(count_per_section)
#plot_histogram(count_per_section, 1, 1)

section  count
17           petota   1734
10       holophylla   1449
13        melongena   1442
20            torva   1332
3      brevantherum    811
19          solanum    577
5         dulcamara    389
8      herposolanum    224
14      micracantha    180
11       lasiocarpa    178
0      acanthophora    100
1   anarrhichomenum    100
12     lycopersicon     76
4          crinitum     66
7     gonatotrichum     58
15      pachyphylla     48
2        basarthrum     42
16      periscariae     15
18        pteroidea     12
6    erythrotrichum      3
9     herpystrichum      2

## Averages and std for sizes and resolution per section

In [ ]:
dedup_images[["section", "filesize_mb", "width","height"]].describe()

filesize_mb        width        height
count  8838.000000  8838.000000   8838.000000
mean      6.191186  3640.765897   5148.799050
std       8.051593  2122.252540   2743.151575
min       0.006400   120.000000    160.000000
25%       1.861500  1216.000000   2000.000000
50%       5.873050  3750.000000   5500.000000
75%       8.962625  6023.250000   8679.250000
max      96.622100  7319.000000  10319.000000

In [ ]:
dedup_images[dedup_images["filesize_mb"] == 96.622100]

Unnamed: 0 section      species  filesize_mb  width  height image_type  \
6029        6087  petota  polyadenium      96.6221   7319   10319        jpg   

     source              hash  \
6029   gbif  dcd8cd4d7b1b9d97   

                                                                                                      full_path  
6029  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_polyadenium_1322105893_gbif_3652.jpeg

In [ ]:
avg_per_section = dedup_images[["section", "filesize_mb", "width","height"]].groupby(
    ["section"], as_index=False)[["filesize_mb", "width", "height"]].mean()
display(avg_per_section)

section  filesize_mb        width       height
0      acanthophora     4.982104  3384.140000  4882.310000
1   anarrhichomenum     3.327075  3623.230000  5183.410000
2        basarthrum     4.919305  3570.309524  5302.952381
3      brevantherum     6.603759  3879.697904  5459.734895
4          crinitum     5.918309  3724.075758  5085.090909
5         dulcamara     5.468920  3634.740360  5208.511568
6    erythrotrichum     9.045133  4754.666667  6655.000000
7     gonatotrichum    10.414095  4834.189655  6540.517241
8      herposolanum     6.437967  3533.379464  4877.651786
9     herpystrichum     9.293550  5271.000000  7256.500000
10       holophylla     6.142634  3643.957902  5177.060041
11       lasiocarpa     7.880952  3767.674157  5330.797753
12     lycopersicon     5.625682  3880.302632  5436.842105
13        melongena     5.926673  3743.023578  5267.512483
14      micracantha     5.948298  4001.950000  5621.422222
15      pachyphylla     5.392396  3561.729167  4951.750000
16      periscariae     5.897307  3558.066667  5229.333333
17           petota     6.245785  3150.388120  4475.539216
18        pteroidea     2.164125  2062.250000  3079.333333
19          solanum     5.853665  3913.372617  5496.374350
20            torva     6.595830  3821.539039  5423.030030

In [ ]:
std_per_section = dedup_images[["section", "filesize_mb", "width","height"]].groupby(
    ["section"], as_index=False)[["filesize_mb", "width", "height"]].std()
display(std_per_section)

section  filesize_mb        width       height
0      acanthophora     3.782009  1946.907701  2537.533451
1   anarrhichomenum     4.120618  2518.173347  3421.364879
2        basarthrum     3.752585  1848.003214  2593.533844
3      brevantherum     7.316670  2072.632264  2670.786588
4          crinitum     4.663881  1979.131964  2781.564594
5         dulcamara     3.927902  1950.253127  2493.599843
6    erythrotrichum     1.889407  1740.133711  2000.518683
7     gonatotrichum    17.774800  2460.044939  3254.423161
8      herposolanum     7.092330  2050.892957  2657.993259
9     herpystrichum     1.835720  2151.018828  2484.066122
10       holophylla     7.725761  2009.480263  2600.346287
11       lasiocarpa    10.882795  2019.433952  2638.686703
12     lycopersicon     3.840170  2311.028296  2933.673347
13        melongena     7.379719  2061.264272  2644.016235
14      micracantha     4.037585  2091.114785  2731.083987
15      pachyphylla     3.287023  1431.492715  1855.302768
16      periscariae     1.635619   743.354604  1048.287492
17           petota    11.685625  2380.328344  3067.366855
18        pteroidea     2.677688  1233.588713  1760.516677
19          solanum     5.650578  1943.453497  2515.065729
20            torva     5.483628  2021.183000  2598.140697

## How many images are smaller than 512x512 and to which section they belong?

Remove them if feasible.

### Width

In [ ]:
#dedup_images.drop("Unnamed: 0", axis=1, inplace=True)
dedup_images[dedup_images["width"] < 512].sort_values("width", ascending=True).head(10)

section         species  filesize_mb  width  height image_type  \
6244        petota    stoloniferum       0.0064    120     160        jpg   
2223    holophylla       nigricans       0.0577    455     768        jpg   
836   brevantherum      umbellatum       0.0440    458     768        jpg   
6119        petota       schenckii       0.0479    461     768        jpg   
2722    holophylla  pseudocapsicum       0.0467    461     768        jpg   
8828         torva          torvum       0.0445    461     768        jpg   
8429         torva     lanceolatum       0.0531    462     768        jpg   
8070         torva     lanceolatum       0.0538    462     768        jpg   
8835         torva          torvum       0.0538    462     768        jpg   
3651     melongena   angustifolium       0.0508    462     768        jpg   

       source              hash  \
6244     gbif  d0b89a9ab2b0f1d8   
2223     gbif  46563625339eb6e8   
836      gbif  98a7b39a9e9ac2e8   
6119  idigbio  9e9a323b3a9579e8   
2722     gbif  ac2c3e96369698e0   
8828  idigbio  ea3ea636938a3ae8   
8429  idigbio  8c4e4c5a4e7cb0c8   
8070     gbif  8e8e16176313cad8   
8835  idigbio  383917333c8eb2c8   
3651  idigbio  d8989a9676b2bad8   

                                                                                                                                           full_path  
6244                                      P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_stoloniferum_1057377186_gbif_4383.jpeg  
2223                                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\holophylla\holophylla_nigricans_1056273794_gbif_4749.jpeg  
836                             P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\brevantherum\brevantherum_umbellatum_1056789629_gbif_4532.jpeg  
6119            P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_schenckii_8145f46f-685f-4c43-b174-f0c766a70e5e_idigbio_3747.jpeg  
2722                            P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\holophylla\holophylla_pseudocapsicum_1056213474_gbif_4770.jpeg  
8828                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva\torva_torvum_b94c0065-67d9-43d8-9b96-e9b68596ff33_idigbio_4724.jpeg  
8429            P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva\torva_lanceolatum_67a09a8c-3bef-4916-b9e0-6148c40f9f63_idigbio_4198.jpeg  
8070                                         P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva\torva_lanceolatum_1057376101_gbif_4386.jpeg  
8835                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\torva\torva_torvum_f782d965-c5c9-4d01-a3ae-440141a62212_idigbio_3164.jpeg  
3651  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\melongena\melongena_angustifolium_2e9167b4-6d90-43f1-bbf9-b8b945dd5040_idigbio_1900.jpeg

### Height

In [ ]:
dedup_images[dedup_images["width"] < 512].sort_values("height", ascending=True).head(10)

section          species  filesize_mb  width  height image_type  \
6244        petota     stoloniferum       0.0064    120     160        jpg   
69    acanthophora     myriacanthum       0.0434    504     768        jpg   
5774        petota       hjertingii       0.0647    500     768        jpg   
5808        petota        iopetalum       0.0425    500     768        jpg   
5899        petota      michoacanum       0.0442    496     768        jpg   
6056        petota      polyadenium       0.0551    493     768        jpg   
6119        petota        schenckii       0.0479    461     768        jpg   
6120        petota        schenckii       0.0417    488     768        jpg   
6131        petota  stenophyllidium       0.0415    488     768        jpg   
6227        petota     stoloniferum       0.0458    500     768        jpg   

       source              hash  \
6244     gbif  d0b89a9ab2b0f1d8   
69    idigbio  d699093e18ca0312   
5774  idigbio  b18fbda12f6b3b63   
5808     gbif  d31b2bdcb0e9f170   
5899     gbif  c40cf28361659549   
6056  idigbio  c94533db8cc6d9d9   
6119  idigbio  9e9a323b3a9579e8   
6120  idigbio  9199333838946569   
6131  idigbio  e36e0c0c183121b1   
6227     gbif  c6c3c3c1c927372f   

                                                                                                                                                full_path  
6244                                           P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_stoloniferum_1057377186_gbif_4383.jpeg  
69    P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_myriacanthum_41c09901-0b87-4be2-8e98-673a7f93ff1d_idigbio_3163.jpeg  
5774                P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_hjertingii_ddc08aec-8bf6-4a16-9cfd-866f935f5cdf_idigbio_3167.jpeg  
5808                                              P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_iopetalum_1057308116_gbif_4403.jpeg  
5899                                            P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_michoacanum_1055440673_gbif_4943.jpeg  
6056                P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_polyadenium_35d4f104-3e7c-4d62-ab61-a44d7d1f537f_idigbio_111.jpeg  
6119                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_schenckii_8145f46f-685f-4c43-b174-f0c766a70e5e_idigbio_3747.jpeg  
6120                 P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_schenckii_8145f46f-685f-4c43-b174-f0c766a70e5e_idigbio_3748.jpeg  
6131           P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_stenophyllidium_0ad7d5a3-157d-4eea-91ee-4a1fd8214d41_idigbio_1622.jpeg  
6227                                           P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\petota\petota_stoloniferum_1055662979_gbif_4902.jpeg

### Removing very small pictures

Given the results of the count, there's only one specimen that is too small to be upscaled to 512x512. The rest of the pictures have values less than 512 but they are close enough that distortion may not be too bad.

In [ ]:
display(dedup_images.shape)
big_pictures = dedup_images.drop(dedup_images[dedup_images["height"] < 200].index)
display(big_pictures.shape)

(8838, 9)

(8837, 9)

In [ ]:
big_pictures.to_csv(os.path.join(root_folder, "Downloaded_dedup_images_nosmall_report.csv"),
                    index=False)
#big_pictures = pd.read_csv(os.path.join(root_folder, "Downloaded_dedup_images_nosmall_report.csv"))


## Remove sections that have less than 100 pictures TBD

In [ ]:
count_per_section = big_pictures.groupby(["section"]).size().reset_index(name='count').sort_values("count", ascending=False)
display(count_per_section)

section  count
17           petota   1733
10       holophylla   1449
13        melongena   1442
20            torva   1332
3      brevantherum    811
19          solanum    577
5         dulcamara    389
8      herposolanum    224
14      micracantha    180
11       lasiocarpa    178
0      acanthophora    100
1   anarrhichomenum    100
12     lycopersicon     76
4          crinitum     66
7     gonatotrichum     58
15      pachyphylla     48
2        basarthrum     42
16      periscariae     15
18        pteroidea     12
6    erythrotrichum      3
9     herpystrichum      2

In [ ]:
sections_of_interest = count_per_section[count_per_section["count"] >= 100]["section"]
display(sections_of_interest)

17             petota
10         holophylla
13          melongena
20              torva
3        brevantherum
19            solanum
5           dulcamara
8        herposolanum
14        micracantha
11         lasiocarpa
0        acanthophora
1     anarrhichomenum
Name: section, dtype: object

In [ ]:
sections_with_100 = big_pictures[big_pictures["section"].isin(sections_of_interest)]
display(sections_with_100.shape)

previous_records = big_pictures.shape[0]
new_records = sections_with_100.shape[0]

print(f"We went from {previous_records} to {new_records}, {previous_records - new_records} were records were removed")

(8515, 9)

We went from 8837 to 8515, 322 were records were removed


In [ ]:
#sections_with_100.to_csv(os.path.join(root_folder, "Downloaded_dedup_images_nosmall_100picspersection.csv"),
#                    index=False)
sections_with_100 = pd.read_csv(os.path.join(root_folder, "Downloaded_dedup_images_nosmall_100picspersection.csv"))

In [15]:
sections_with_100.head(5)

section      species  filesize_mb  width  height image_type source  \
0  acanthophora  acerifolium       8.9974   3750    5500        jpg   gbif   
1  acanthophora  acerifolium       7.3703   3750    5500        jpg   gbif   
2  acanthophora  acerifolium       7.0359   3750    5500        jpg   gbif   
3  acanthophora  acerifolium       7.8730   3750    5500        jpg   gbif   
4  acanthophora  acerifolium       5.9595   3750    5500        jpg   gbif   

               hash  \
0  8c86b6326a6e0cca   
1  bea48d0a2aac92b2   
2  b893b2a893b6a2e6   
3  98a6963cb69a8aca   
4  96963696ca8adaf2   

                                                                                                              full_path  
0  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928496814_gbif_2700.jpg  
1  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928638909_gbif_2609.jpg  
2  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1928803178_gbif_2510.jpg  
3  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1929578053_gbif_2035.jpg  
4  P:/CODE/ITESM/tesis-dataset-downloader/solanum_output\acanthophora\acanthophora_acerifolium_1929578350_gbif_2034.jpg

# 4 Resizing of relevant images to build final pre-processed dataset TBD

In [40]:
def win2linux_path(path: str) -> str:
    path = path.replace("\\", "/")
    return path.replace("P:", "/mnt/pepper")

def linux2win_path(path: str) -> str:
    path = path.replace("\\", "/")
    return path.replace("/mnt/pepper", "P:")

def get_filename(path: str) -> str:
    return path.split("/")[-1]

def create_folders(path: str):
    if not os.path.exists(path):
        os.makedirs(path)


In [44]:
%%time
# Output to a folder called resized that respects each of the folders for each section
resized_folder = os.path.join(root_folder, "z_clean_resized")
create_folders(resized_folder)

# Build a new dataframe with the new images, size in MB and save it to csv
# Print the reduction in size for the dataset
new_images = []

# For each of the records, read the input image and resize it
for index, row in sections_with_100.iterrows():
    
    # Change size
    image_path = win2linux_path(row["full_path"])
    image = Image.open(image_path)
    image = image.resize((512, 512))

    # Output folder
    output_folder = os.path.join(resized_folder, row["section"])
    create_folders(output_folder)

    output_file = os.path.join(output_folder, get_filename(image_path))
    image.save(output_file)

    # Size (MB)
    filesize_mb = round(os.path.getsize(output_file) / 1024 / 1024, 4)
    
    row = {
        "section": row["section"],
        "species": row["species"],
        "filesize_mb": filesize_mb,
        "image_type": row["image_type"],
        "source": row["source"],
        "full_path": linux2win_path(output_file),
    }

    new_images.append(row)

resized_images = pd.DataFrame(new_images)

/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/linux-venv/lib/python3.9/site-packages/PIL/TiffImagePlugin.py:822: UserWarning: Corrupt EXIF data.  Expecting to read 12 bytes but only got 10. 
  warnings.warn(str(msg))


CPU times: user 51min 36s, sys: 2min 48s, total: 54min 24s
Wall time: 59min 32s


In [45]:
%%time
#resized_images.to_csv(os.path.join(resized_folder, "images_dedup_512x512_100picspersection.csv"), index=False)
resized_images = pd.read_csv(os.path.join(resized_folder, "images_dedup_512x512_100picspersection.csv")

CPU times: user 126 ms, sys: 16.6 ms, total: 143 ms
Wall time: 533 ms


In [51]:
size_before = sections_with_100["filesize_mb"].sum()
size_after = resized_images["filesize_mb"].sum()

print(f"Dataset was resized from {size_before} MB to {size_after} MB, a reduction of {1 - (size_after / size_before)}%")
print(f"Before shape: {sections_with_100.shape}\nAfter shape: {resized_images.shape}")


Dataset was resized from 52669.9167 MB to 251.24760000000003 MB, a reduction of 0.9952297703178254%
Before shape: (8515, 9)
After shape: (8515, 6)


## Zip it up in a file so that the dataset can be uploaded to the DGX-1

In [ ]:
!ls '/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/solanum_output/z_clean_resized'
!zip -r '/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/solanum_output/clean_resized_dataset.zip' "/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/solanum_output/z_clean_resized"

In [ ]:
!ls -lh '/mnt/pepper/CODE/ITESM/tesis-dataset-downloader/solanum_output/'